### PR TITLE
openssl3: init at 3.0.0

### PR DESCRIPTION
**Pull request by baloo**
###### Motivation for this change

openssl 3.0.0 released this morning.

I did not change the default version of openssl just yet, this would need to be targeted to staging.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by baloo** [comment]
openssl3 brings a couple cool features, amongst those, KTLS support which I'm excited about.

**Comment by risicle** [comment]
Builds happily, macos 10.15 & nixos x86_64

File: pkgs/development/libraries/openssl/default.nix [hunk 1]
```diff
@@ -205,4 +205,16 @@ in {
...
     withDocs = true;
   };
+
+  openssl_3_0 = common {
+    version = "3.0.0";
+    sha256 = "sha256-We7fy0bCUhTJvTftYHgpe03wHQEiZ/6enu4x9hvHBTY=";
+    patches = [
+      ./3.0/nix-ssl-cert-file.patch
+      (if stdenv.hostPlatform.isDarwin
+       then ./1.1/use-etc-ssl-certs-darwin.patch
```
Comment (andir) [inline pkgs/development/libraries/openssl/default.nix:215]: Shouldn't these patches come from the 3.0 folder or are they the same and the ones in this PR aren't required anymore?

File: pkgs/development/libraries/openssl/default.nix [hunk 2]
```diff
@@ -205,4 +205,16 @@ in {
...
+       else ./1.1/use-etc-ssl-certs.patch)
```
Comment (mweinelt) [inline pkgs/development/libraries/openssl/default.nix:216]: I'd rather we duplicate the patches into a `3.0` directory, since this looks very error-prone.

**Comment by SuperSandro2000** [comment]
> this would need to be targeted to staging.

Far more importing is how many breakages it would cause.

File: pkgs/development/libraries/openssl/default.nix (hunk 2 above)
Comment (SuperSandro2000) [reply pkgs/development/libraries/openssl/default.nix:216]: Or we move them up one directory.

File: pkgs/development/libraries/openssl/default.nix (hunk 1 above)
Comment (baloo) [reply pkgs/development/libraries/openssl/default.nix:215]: They're the same, I just copied them and forgot to update the reference.

**Comment by baloo** [comment]
> > this would need to be targeted to staging.
> 
> Far more importing is how many breakages it would cause.

Yeah, that's for sure. I just included here like that since I want to start playing with it.

File: pkgs/development/libraries/openssl/default.nix [hunk 3]
```diff
@@ -198,11 +198,29 @@ in {
...
+
+  openssl_3_0 = common {
+    version = "3.0.0";
+    sha256 = "sha256-We7fy0bCUhTJvTftYHgpe03wHQEiZ/6enu4x9hvHBTY=";
+    patches = [
+      ./3.0/nix-ssl-cert-file.patch
+
+      # openssl will only compile in KTLS if the current kernel supports it.
+      # This patch disables build-time detection.
+      ./3.0/openssl-disable-kernel-detection.patch
```
Comment (baloo) [inline pkgs/development/libraries/openssl/default.nix:217]: openssl builds will only enable KTLS if the builder has kernel-headers available, and if CONFIG_KTLS is enabled there. I believe this breaks the isolation and separation of concerns.

Disabling detection and building ktls none-the-less will throw runtime error when user will want to do KTLS-related operation, and should fallback to the usual implementation.

**Comment by baloo** [comment]
up

**Review by ajs124 — APPROVED** [review]

**Comment by dasJ** [comment]
Any objections left?

**Pull request merged**